gstd/src/common/reservation_manager.rs

```diff
@@ -0,0 +1,155 @@
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use super::primitives::ReservationId;
+use crate::{errors::Result, exec, prelude::*};
+
+struct Reservation {
+    id: ReservationId,
+    amount: u64,
+    valid_until: u32,
+}
+
+/// Reservation manager.
+///
+/// The manager is used to control multiple gas reservations
+/// across executions.
+///
+/// # Examples
+///
+/// Create gas reservations inside `init` and use them inside `handle`
+///
+/// ```
+/// use gstd::ReservationId;
+///
+/// static mut RESERVATION_MANAGER: Option<ReservationManager> = None;
+///
+/// extern "C" fn init() {
+///     unsafe {
+///         RESERVATION_MANAGER = Some(ReservationManager::new());
+///
+///         RESERVATION_MANAGER.as_mut().unwrap().reserve(200_000, 50);
+///         RESERVATION_MANAGER.as_mut().unwrap().reserve(100_000, 100);
+///         RESERVATION_MANAGER.as_mut().unwrap().reserve(50_000, 30);
+///     }
+/// }
+///
+/// extern "C" fn handle() {
+///     let reservation = unsafe {
+///         RESERVATION_MANAGER
+///             .as_mut()
+///             .unwrap()
+///             .try_take_reservation(100_000)
+///     };
+///     match reservation {
+///         Some((rid, amount)) => {
+///             msg::send_bytes_from_reservation(
+///                 rid,
+///                 msg::source(),
+///                 b"send_bytes_from_reservation",
+///                 0,
+///             )
+///             .expect("Failed to send message from reservation");
+///         }
+///         None => {
+///             msg::send_bytes(msg::source(), b"send_bytes", 0).expect("Failed to send message");
+///         }
+///     }
+/// }
+/// ```
+///
+/// # See also
+/// - [`ReservationId`](ReservationId) is used to reserve and unreserve gas
+/// for program execution later.
+#[derive(Default)]
+pub struct ReservationManager {
+    reservations: Vec<Reservation>,
+}
+
+impl ReservationManager {
+    /// Create a new ReservationManager
+    pub fn new() -> ReservationManager {
+        ReservationManager {
+            reservations: vec![],
+        }
+    }
+
+    /// Reserve the `amount` of gas for further usage.
+    ///
+    /// `duration` is the block count within which the reserve must be used.
+    pub fn reserve(&mut self, amount: u64, duration: u32) -> Result {
+        let id = ReservationId::reserve(amount, duration)?;
+        let block_height = exec::block_height();
+        let valid_until = duration.saturating_add(block_height);
+
+        let insert_to = self.reservations.partition_point(|reservation| {
+            reservation.amount < amount && reservation.valid_until < valid_until
```

Comment:
what's the property of the layout of reservations? I guess we can't say that `reserve` stores reservations in the sorted order.